### ISSUE TITLE
Switch Hex helper from message passing over `stdio` to using either the new Hex CLI or IPC 

### ISSUE THREAD

**Issue by jeffwidman** [dependabot/dependabot-core#7945]
### Code improvement description

Described in detail here:
* https://github.com/dependabot/dependabot-core/pull/6026#issuecomment-1336276793
* https://github.com/dependabot/dependabot-core/pull/6026#issuecomment-1336310884
* https://github.com/dependabot/dependabot-core/pull/6026#issuecomment-1336473731

Also related:
* https://github.com/dependabot/dependabot-core/issues/2757

Filing this ticket mostly so we don't lose track of it as a source of potential improvements.

**Comment by jeffwidman** [comment]
## Investigation and implementation handoff (2026-08-06)

I investigated the current Hex helper end to end, the historical failures linked above, the related fixes, and the current Hex/Mix CLI surface. This is intended to be enough context for another engineer or coding agent to start implementation without repeating that research.

### Recommendation

Proceed with a **CLI-first refactor using filesystem artifacts**, not a new stdin/stdout protocol and not socket IPC as the first choice.

The target should be:

1. Use official `mix` / `mix hex.*` commands for dependency updates and credential configuration where they preserve current behavior.
2. Read generated files (`mix.lock` and a JSON inspection artifact) rather than sending encoded values through stdout.
3. Keep stdout/stderr as human-readable diagnostics only.
4. Retain one small, named Dependabot Mix task for structured dependency inspection, because neither current Mix nor Hex exposes all of the metadata Dependabot needs in a machine-readable format.
5. Delete the existing dispatcher and the update-specific Elixir scripts once parity is demonstrated.

A socket or `Mix.Shell.Process` IPC design remains a fallback if the CLI spikes below uncover a blocker. It should not be the default: it would replace one custom protocol with another even though all three operations already happen in disposable directories and can communicate through files.

### What exists today

The current flow is substantially more complicated than the Ruby call sites suggest:

```text
Ruby
  -> JSON request on stdin
  -> mix run hex/helpers/lib/run.exs
  -> System.cmd("mix", ["run", ...operation_script])
  -> private Mix/Hex APIs
  -> Erlang term
  -> term_to_binary
  -> Base64 on child stdout
  -> binary_to_term in run.exs
  -> JSON on parent stdout
  -> Ruby JSON parsing
```

Relevant files:

- [`hex/helpers/lib/run.exs`](https://github.com/dependabot/dependabot-core/blob/main/hex/helpers/lib/run.exs): stdin dispatcher, nested Mix process, credential configuration, Base64/Erlang decoding, JSON response.
- [`hex/helpers/lib/check_update.exs`](https://github.com/dependabot/dependabot-core/blob/main/hex/helpers/lib/check_update.exs): calls `Mix.Dep.Fetcher.by_name/4`, applies a 30-second timeout, reads the updated lock, returns an encoded version.
- [`hex/helpers/lib/do_update.exs`](https://github.com/dependabot/dependabot-core/blob/main/hex/helpers/lib/do_update.exs): calls the same private fetcher, runs `deps.get`, reads `mix.lock`, returns its contents as an encoded Erlang term.
- [`hex/helpers/lib/parse_deps.exs`](https://github.com/dependabot/dependabot-core/blob/main/hex/helpers/lib/parse_deps.exs): calls `Mix.Dep.Converger.converge/1` and emits Dependabot-specific metadata for Hex, Git, path, and umbrella dependencies.
- [`hex/lib/dependabot/hex/file_parser.rb`](https://github.com/dependabot/dependabot-core/blob/main/hex/lib/dependabot/hex/file_parser.rb): merges stderr into stdout and contains recovery logic that scans noisy output for a JSON result.
- [`hex/lib/dependabot/hex/update_checker/version_resolver.rb`](https://github.com/dependabot/dependabot-core/blob/main/hex/lib/dependabot/hex/update_checker/version_resolver.rb): has a second recovery path that tries to parse the last output line as JSON.
- [`hex/lib/dependabot/hex/file_updater/lockfile_updater.rb`](https://github.com/dependabot/dependabot-core/blob/main/hex/lib/dependabot/hex/file_updater/lockfile_updater.rb): receives the entire lockfile through the helper protocol.
- [`hex/lib/dependabot/hex/credential_helpers.rb`](https://github.com/dependabot/dependabot-core/blob/main/hex/lib/dependabot/hex/credential_helpers.rb): flattens credentials into positional string arrays.

The child scripts now redirect the standard Erlang logger to stderr, thanks to [#12242](https://github.com/dependabot/dependabot-core/pull/12242), but the architecture still assumes that every other writer will leave stdout untouched. Warnings from evaluated `mix.exs` files, Hex solver diagnostics, libraries, and future runtime changes can violate that assumption.

This is the root of errors such as:

```text
ArgumentError: invalid external representation of a term
```

The historical analysis in [the first linked PR comment](https://github.com/dependabot/dependabot-core/pull/6026#issuecomment-1336276793) showed Hex 2's human-readable conflict explanation appearing before the encoded result. #12242 fixed another occurrence where `Hex.Registry.Server` logs contaminated the response. Those were valid local fixes, but both are symptoms of stdout being used simultaneously as a log stream and an unframed data transport.

There is a second, independent maintenance problem: the helper directly calls private APIs:

- `Mix.Dep.Fetcher.by_name/4`
- `Mix.Dep.Converger.converge/1`
- `Mix.Dep.Lock.read/0`
- `Hex.Config.read/0`, `read_repos/1`, and `update_repos/1`
- `Hex.Repo.get_repo/1` and `get_public_key/1`

A recent tuple-order change around `Hex.Repo.get_public_key/1` caused a real public-key decode regression. The current tests now characterize that error, but the better boundary is Hex's supported CLI.

### Current tool capabilities

The image currently pins:

- Erlang/OTP 27
- Elixir/Mix 1.19.5
- Hex 2.3.1

As of this investigation, the latest stable releases are Hex 2.5.1 and Elixir/Mix 1.20.3.

#### Hex 2.5.1

Hex 2.5.0 added:

```shell
mix hex.outdated --json
```

OTP 27 is already sufficient for this feature. The output is an array containing:

```json
{
  "package": "plug",
  "only": "",
  "lock_version": "1.16.0",
  "latest_version": "1.18.1",
  "requirements": [
    {
      "source": "mix.exs",
      "requirement": "~> 1.0",
      "up_to_date": true
    }
  ],
  "outdated": true,
  "cooldown": null
}
```

This is useful for diagnostics and candidate metadata, but **must not replace actual dependency resolution**. Hex's own task documentation says it compares the lock and current requirements without performing the full solver operation. Updating one dependency can update children and introduce constraints that make the reported latest version impossible. Dependabot must continue to use the real solver result.

Moving from Hex 2.3.1 to 2.5.1 is independently worthwhile. The intervening releases include:

- CVE-2026-21619: safe deserialization of Erlang terms in API responses (2.3.2).
- OAuth and private-repository authentication improvements (2.4.x).
- CVE-2026-32148: fail on `mix.lock` checksum mismatch instead of silently rewriting it (2.4.2).
- Private package timeout/auth fixes (2.4.2).
- JSON output, dependency policies, release cooldowns, and security-advisory warnings (2.5.0+).

Hex 2.5.1 is compatible with the currently pinned Elixir 1.19.5, so the Hex bump does not need to wait for an Elixir upgrade.

#### Mix 1.20.3

Mix 1.20 added dependency-name filtering to the human-readable command:

```shell
mix deps plug phoenix
```

It did **not** add JSON dependency listing, a lockfile JSON command, or an output-file option. Its `mix deps.update NAME` implementation still calls `Mix.Dep.Fetcher.by_name/4` internally, but that call is now behind a supported CLI contract.

Elixir 1.20.3 therefore does not eliminate the custom inspector. Upgrade it independently if desired; it is not a prerequisite for this refactor.

### Proposed operation boundaries

#### 1. Update a dependency / generate a lockfile

In the existing disposable project directory, run the supported command:

```shell
mix deps.update DEPENDENCY --no-archives-check
```

Then read `mix.lock` directly from disk.

This is the public equivalent of the private `Mix.Dep.Fetcher.by_name/4` call currently used by both update scripts: it unlocks the named dependency and its children and resolves/fetches them according to the project requirements. The updater should preserve the existing Ruby-side mixfile preparation and the existing lockfile formatting post-processing.

`do_update.exs` should become unnecessary. `mix.lock` is already the authoritative artifact; sending it through Erlang-term and JSON encoders adds no value.

#### 2. Find the latest resolvable version

Use the same solver operation in a disposable copy:

```shell
mix deps.update DEPENDENCY --no-archives-check
```

Then inspect the resulting lock through the structured inspector described below and return the selected Hex version or Git SHA to Ruby.

Preserve the current two-pass behavior in `VersionResolver`:

1. Resolve with the prepared/relaxed requirements.
2. If that fails, resolve the original files to distinguish “this update is not possible” from “the user's original dependency files are not resolvable.”

Also preserve:

- The dedicated resolution timeout (currently 30 seconds).
- `Dependencies have diverged` behavior.
- Git dependency results represented by a 40-character SHA.
- Current mappings to `PrivateSourceAuthenticationFailure`, `DependencyFileNotResolvable`, and other Dependabot errors.

Do not use `hex.outdated --json` as the selected version because it is not a solver.

#### 3. Parse dependency metadata

There is no complete standard CLI replacement for `parse_deps.exs`.

Dependabot currently needs fields that `mix deps` and `hex.outdated --json` do not jointly provide, including:

- application name versus Hex package name (`:hex` aliasing),
- requirement,
- source mixfile,
- top-level status,
- `:only` groups,
- lock checksum,
- Git URL/branch/ref/SHA,
- path and umbrella expansion semantics,
- unlocked dependencies.

Convert the existing parsing logic into one controlled command, for example:

```shell
mix dependabot.hex.inspect --output /absolute/path/dependencies.json
```

Recommended contract:

- Evaluate the sanitized project in the same way the current parser does.
- Write versioned JSON to the requested path.
- Write to a sibling temporary file and rename atomically.
- Return nonzero on failure.
- Never put the result on stdout.
- Validate artifact existence, schema version, and expected types in Ruby.
- Let stdout/stderr contain arbitrary warnings and diagnostics.

The implementation may still need `Mix.Dep.Converger.converge/1`; if so, isolate that private API dependency in this one task instead of spreading private calls across three scripts. A future upstream `mix deps --json --output PATH` could replace it.

Packaging needs a small spike. The cleanest user-facing shape is likely a precompiled Mix archive/task installed into the helper `MIX_HOME`. A precompiled module invoked through a controlled `mix run` entry point is also acceptable if archive task discovery causes compatibility problems. The important contract is the explicit output file, not which of those two packaging mechanisms wins.

#### 4. Configure private repositories

Create a fresh, operation-scoped `HEX_HOME` and configure only the credentials supplied for that job. Keep the preinstalled Hex archive in the existing `MIX_HOME`; do not share repository credentials or cache configuration between jobs.

For Hex organizations, use the supported noninteractive CI command:

```shell
mix hex.organization auth ORGANIZATION --key TOKEN
```

For custom repositories with a supplied fingerprint, use:

```shell
mix hex.repo add NAME URL \
  --auth-key TOKEN \
  --fetch-public-key FINGERPRINT
```

This moves organization-key verification and fingerprint handling behind Hex's public CLI and removes `Hex.Config` / `Hex.Repo` calls from `run.exs`.

There is one important parity gap to resolve before deleting the current code: Dependabot currently downloads and stores a custom repository's public key even when `public_key_fingerprint` is empty. Hex 2.5.1's `--fetch-public-key` option requires a fingerprint; omitting it adds the repository without fetching a key. Do not silently reduce registry verification. The implementation spike should choose and test one of:

1. Fetch the PEM with an authenticated, structured HTTP client and pass a temporary file via `mix hex.repo add ... --public-key PATH`.
2. Retain a very small compatibility path for no-fingerprint credentials until that behavior can be deprecated.
3. Require fingerprints only after a deliberate credential-contract/security review and migration plan.

Option 1 best preserves behavior without using private Hex APIs.

### CLI runner requirements

Add one Ruby abstraction responsible for all Hex/Mix command execution. It should:

- Accept argv as an array rather than interpolating a shell string.
- Set operation-scoped `HEX_HOME` and the existing controlled Mix environment.
- Apply explicit timeouts and kill the process tree on timeout.
- Capture stdout and stderr as diagnostics without using either as the result channel.
- Redact auth keys/tokens in command descriptions, exceptions, metrics, and debug output.
- Preserve exit status and both output streams in error context.
- Validate expected artifact paths after successful exit.
- Clean the credential home and result files when the temporary operation ends.

Hex's official credential commands require keys as command arguments. Avoid `SharedHelpers.run_shell_command` with a fully interpolated command, since it may log the literal token. Use an argv-based subprocess API with explicit redaction. Process-table visibility is still possible during the short-lived CLI invocation; document and assess that within the existing single-purpose updater container threat model.

Do not set `stderr_to_stdout: true` for result parsing. Combining them is fine only when constructing a human-readable exception after the process has exited.

### Two compatibility spikes to run first

Before the broad rewrite, prove these against fixtures in the Hex development image:

1. **Mix aliases:** verify whether a project-defined alias named `deps.update` can intercept `mix deps.update DEPENDENCY`. The current direct fetcher call bypasses aliases. If aliases can intercept it, invoke the official task from a controlled Dependabot task/module in a way that bypasses project aliases, while retaining the normal Mix/Hex solver.
2. **Custom repository without a fingerprint:** establish the safe public-key-fetch flow described above and verify the same authentication/fingerprint errors Ruby currently maps.

Also prove the command against missing lockfiles, Git dependencies, umbrella projects, and a `mix.exs` that prints warnings. Those are cheap discriminating checks for whether the CLI boundary is viable.

### Suggested PR sequence

Keeping these changes separate should make regressions easier to diagnose and review:

1. **Characterization and Hex bump**
   - Add/strengthen end-to-end tests for noisy stdout/stderr, timeout, missing lock, aliases, custom repos, organizations, fingerprints, Git deps, and umbrellas.
   - Upgrade Hex 2.3.1 to 2.5.1 in `hex/Dockerfile` and update its checksum/build inputs as needed.
   - Do not change the helper protocol in this PR.

2. **Shared CLI runner and credential home**
   - Add argv execution, redaction, timeout, diagnostics, and operation-scoped `HEX_HOME`.
   - Configure organizations/custom repos through official CLI commands.
   - Resolve the no-fingerprint parity gap.
   - Keep existing operation scripts temporarily so credential behavior can be validated independently.

3. **Lockfile update via CLI**
   - Replace `do_update.exs` with `mix deps.update NAME` plus reading `mix.lock`.
   - Preserve lock formatting and error classification.
   - Delete `do_update.exs` after focused and ecosystem tests pass.

4. **Version resolution via CLI**
   - Replace `check_update.exs` with the solver command plus structured lock inspection.
   - Preserve timeout and two-pass original/prepared semantics.
   - Delete `check_update.exs` after parity tests pass.

5. **Output-file inspector**
   - Convert `parse_deps.exs` into the named/precompiled task with `--output`.
   - Change `FileParser` to read and validate the JSON artifact.
   - Remove all “find JSON in noisy output” recovery code.

6. **Protocol deletion and cleanup**
   - Delete `run.exs` and `CredentialHelpers` positional serialization.
   - Remove Erlang term encoding, Base64, stdin requests, and JSON stdout responses.
   - Remove stale comments/workarounds that assume warnings corrupt results.
   - Update helper build/package documentation.

An upstream follow-up to Mix proposing a structured command such as `mix deps --json --output PATH` would be useful, but this refactor should not wait on it.

### Acceptance test matrix

At minimum, cover:

- Public Hex dependency: update available, no update, and unresolvable update.
- Missing and existing `mix.lock`.
- Exact, range, regex, bare, aliased-package, and prerelease requirements.
- Git dependency by branch, tag, ref, and returned SHA.
- Path dependencies and umbrella children.
- `:only` groups and divergent environments.
- `mix.exs` reading/evaluating supporting files.
- Invalid or raising `mix.exs`.
- Arbitrary stdout, logger output, compiler warnings, solver conflict output, and stderr.
- Resolution timeout and process cleanup.
- Organization credentials: valid, missing, and invalid.
- Custom repository auth: valid/invalid key, unreachable key endpoint, decodable/invalid PEM, matching/mismatching/empty fingerprint.
- Multiple private repositories in one job and no credential leakage into the next job.
- Lockfile checksum enforcement under Hex 2.5.1.
- Existing compact versus multiline lockfile formatting.
- User/project Hex policies and cooldown configuration, ensuring the real solver remains authoritative.
- Project aliases around `deps.update`.

Useful regression anchors already in the suite include the warning-producing resolver fixture and the private repository/fingerprint cases. During this investigation, the focused warning test passed in the development container:

```text
bin/test hex spec/dependabot/hex/update_checker/version_resolver_spec.rb:118
1 example, 0 failures
```

### Non-goals / cautions

- This does not make evaluating `mix.exs` a security sandbox. Dependabot already evaluates sanitized project code in a container; removing the nested process does not materially change that trust boundary. Preserve the existing sanitizer and container isolation.
- Do not parse `mix deps` human-readable output. That would recreate the same class of compatibility problem in a different format.
- Do not treat `hex.outdated --json` as a solver.
- Do not pass secrets through JSON/stdin merely to avoid CLI argv unless there is a supported consumer; inventing another credential protocol would undermine this simplification.
- Do not switch to an unreleased Hex build for this work. Hex 2.5.1 provides the useful released functionality; later features can be adopted after release.

### Sources checked

- Original failure analysis: https://github.com/dependabot/dependabot-core/pull/6026#issuecomment-1336276793
- Original CLI/IPC discussion: https://github.com/dependabot/dependabot-core/pull/6026#issuecomment-1336310884 and https://github.com/dependabot/dependabot-core/pull/6026#issuecomment-1336473731
- Logging mitigation: https://github.com/dependabot/dependabot-core/pull/12242
- Hex 2.5.1 changelog: https://github.com/hexpm/hex/blob/v2.5.1/CHANGELOG.md
- `hex.outdated --json` implementation and its solver caveat: https://github.com/hexpm/hex/blob/v2.5.1/lib/mix/tasks/hex.outdated.ex
- Hex repository CLI implementation: https://github.com/hexpm/hex/blob/v2.5.1/lib/mix/tasks/hex.repo.ex
- Hex organization CLI implementation: https://github.com/hexpm/hex/blob/v2.5.1/lib/mix/tasks/hex.organization.ex
- Mix 1.20.3 `deps.update`: https://github.com/elixir-lang/elixir/blob/v1.20.3/lib/mix/lib/mix/tasks/deps.update.ex
- Mix 1.20.3 `deps`: https://github.com/elixir-lang/elixir/blob/v1.20.3/lib/mix/lib/mix/tasks/deps.ex

In short: newer Hex/Mix releases reduce the custom helper surface but do not eliminate it. The practical endpoint is **official CLI for credentials and solving, files for results, and one explicit-output Dependabot inspector for metadata**. That removes the fragile transport and most private API usage while retaining full solver semantics.

**Comment by jeffwidman** [comment]
### Follow-up corrections and additional implementation notes

A few corrections and gaps surfaced from additional source review and fixture probes:

#### Upgrade Elixir/Mix as part of the toolchain bump

Elixir 1.19.5 is affected by [GHSA-w2h8-8x3g-278p](https://github.com/elixir-lang/elixir/security/advisories/GHSA-w2h8-8x3g-278p), an unbounded integer parsing DoS in `Version`, fixed in 1.20.1+. Dependabot evaluates attacker-controlled manifests and package metadata, so this is relevant to the updater threat model. I recommend pinning Elixir/Mix 1.20.3 together with Hex 2.5.1 in the first PR while retaining OTP 27. The published SHA-256 for the OTP 27 Elixir archive is:

```
28b1707b69df18515f63242e8d218d8dffd31dadfc21bfe649618ca1f1b41b73
```

#### Two CLI-output corrections

A real Hex 2.5.1/OTP 27 probe produced `"cooldown":"nil"`, not JSON `null`, from `mix hex.outdated --json`. It also returned exit status 1 while emitting valid JSON because outdated packages were present. This does not affect the recommendation not to use it as the solver, but any diagnostic consumer must accept that status and schema behavior.

Also, Mix 1.20 did add `mix deps.tree --format dot --output PATH`. It still did not add a useful structured dependency inventory or lockfile JSON command, so it does not remove the need for the inspector; the earlier statement that Mix added no output-file option was simply too broad.

#### CLI behavior already validated

The following no longer need to be treated as open viability spikes:

- A public Hex dependency updates correctly through `mix deps.update NAME`.
- Required transitive children update with it, preserving the current `Mix.Dep.Fetcher.by_name/4` semantics.
- A Git dependency updates and yields the expected lock SHA.
- An unresolvable update exits nonzero with useful solver diagnostics and leaves `mix.lock` unchanged.
- Hex 2.5.1 runs on the current Elixir 1.19.5/OTP 27 image.

Missing-lockfile and umbrella behavior, aliases, and private-repository parity remain worth explicit characterization.

#### Alias bypass must happen below normal task dispatch

Mix source confirms that `Mix.Task.run/2` gives project aliases precedence. A project can also alias a new name such as `dependabot.hex.update`, so merely introducing a namespaced task does not guarantee bypass. The update entry point should load the sanitized project and invoke the intended task module through a controlled path, for example `Mix.Tasks.Deps.Update.run/1`, rather than dispatching its name through `Mix.Task.run/2`. Keep `--no-archives-check` in the argument list because `Mix.Tasks.Deps.Update.run/1` handles it directly.

#### `HEX_HOME` is not complete environment isolation

Hex environment variables override configuration even with a fresh `HEX_HOME`. The runner needs an explicit inherited-environment policy for values such as `HEX_REPOS_KEY`, `HEX_API_KEY`, `HEX_API_URL`, unsafe verification flags, mirrors, offline mode, cooldown, and policy. Scrub ambient credentials and unsafe overrides while deliberately preserving intended proxy, CA, and policy inputs.

`HEX_HOME` also contains cache/data as well as credentials. Creating a fresh home for every operation may repeatedly download registry state. Benchmark operation-scoped versus job-scoped homes, or seed each operation from a credential-free baseline cache. Whichever scope wins, create secret-bearing directories/files with explicit `0700`/`0600` permissions.

#### Bound diagnostics and runtime independently

The current subprocess capture accumulates stdout/stderr without a size limit and uses an inactivity timeout that resets whenever output arrives. A noisy `mix.exs` can therefore avoid the idle timeout while consuming memory. The new runner should have an absolute wall-clock deadline, bounded/truncated diagnostic buffers, and a maximum accepted artifact size in addition to process-tree cleanup.

#### Credential CLI parity needs two extra checks

`mix hex.organization auth ORGANIZATION --key KEY` verifies the key through the Hex API before storing it. That adds an API request not currently made during credential setup, so test the case where the package repository is reachable but the auth API is unavailable and preserve useful error classification.

For a custom repository without a fingerprint, a generic HTTP fetch of `/public_key` is not automatically equivalent to `Hex.Repo.get_public_key/1`: Hex applies repository URL construction, authentication/OAuth behavior, proxy and CA settings, and Hex-specific response handling. Retaining a tiny Hex-side compatibility adapter may be safer than reproducing those semantics in Ruby until Hex exposes a supported no-fingerprint fetch command.
